contracts/eip721/EIP721.sol

```diff
@@ -0,0 +1,311 @@
+        return approvedOperators[_owner][_operator];
+    }
+
+    /// @notice A distinct Uniform Resource Identifier (URI) for a given asset.
+    /// @dev Throws if `_tokenId` is not a valid NFT. URIs are defined in RFC
+    ///  3986. The URI may point to a JSON file that conforms to the "ERC721
+    ///  Metadata JSON Schema".
+    function tokenURI(uint256 _tokenId) external view returns (string) {
+        return tokenURIs[_tokenId];
+    }
+
+    /// @notice Query if a contract implements an interface
+    /// @param interfaceID The interface identifier, as specified in ERC-165
+    /// @dev Interface identification is specified in ERC-165. This function
+    ///  uses less than 30,000 gas.
+    /// @return `true` if the contract implements `interfaceID` and
+    /// `interfaceID` is not 0xffffffff, `false` otherwise
+    function supportsInterface(bytes4 interfaceID) external view returns (bool) {
+
+        if (interfaceID == ERC721_BASE_INTERFACE_SIGNATURE ||
+        interfaceID == ERC721_METADATA_INTERFACE_SIGNATURE ||
+        interfaceID == ERC721_ENUMERABLE_INTERFACE_SIGNATURE) {
+            return true;
+        } else { return false; }
+    }
+
+    /* -- Internal Functions -- */
+    function checkIfAllowedToOperate(uint256 _tokenId) internal view returns (bool) {
+        return ownerOfToken[_tokenId] == msg.sender || approvedOperators[ownerOfToken[_tokenId]][msg.sender];
+    }
+
+    function internalApprove(address _owner, address _approved, uint256 _tokenId) internal {
+        require(_approved != _owner); //can't approve to owner to itself
+
+        // Note: the following code is equivalent to: require(getApproved(_tokenId) != 0) || _approved != 0);
+        // However: I found the following easier to read & understand.
+        if (approvedOwnerOfToken[_tokenId] == 0 && _approved == 0) {
+            revert(); // add reason for revert?
+        } else {
+            approvedOwnerOfToken[_tokenId] = _approved;
+            emit Approval(_owner, _approved, _tokenId);
+        }
+    }
+
+    function settleTransfer(address _from, address _to, uint256 _tokenId) internal {
+        //clear pending approvals if there are any
+        if (approvedOwnerOfToken[_tokenId] != 0) {
+            internalApprove(_from, 0, _tokenId);
+        }
+
+        removeToken(_from, _tokenId);
+        addToken(_to, _tokenId);
+
+        emit Transfer(_from, _to, _tokenId);
+    }
+
+    function addToken(address _to, uint256 _tokenId) internal {
+        allTokens.push(_tokenId);
+        allTokensIndex[_tokenId] = allTokens.length-1;
+
+        // set token to be owned by address _to
+        ownerOfToken[_tokenId] = _to;
+        // add that token to an array keeping track of tokens owned by that address
+        ownedTokens[_to].push(_tokenId);
+        // shorten length
+        ownedTokensIndex[_tokenId] = ownedTokens[_to].length-1;
+    }
+
+    function removeToken(address _from, uint256 _tokenId) internal {
```

Comment:
`_owner` would be a better name for this parameter, since it is used to get the owner of the token.